appserver/extras/certificate-management/src/main/java/fish/payara/certificate/management/CertificateManagementUtils.java

```diff
@@ -0,0 +1,319 @@
+        expandedKeytoolCmd[i + 1] = "SAN=";
+
+        for (String altName : alternativeNames) {
+            // Check if the altname was provided without any additional info, assuming it's DNS if so
+            if (!altName.contains(",") && !altName.contains(":")) {
+                expandedKeytoolCmd[i + 1] += "DNS:" + altName;
+            } else {
+                expandedKeytoolCmd[i + 1] += altName;
+            }
+            expandedKeytoolCmd[i + 1] += ",";
+        }
+
+        // Remove trailing comma
+        expandedKeytoolCmd[i + 1] = expandedKeytoolCmd[i + 1].substring(0, expandedKeytoolCmd[i + 1].length() - 1);
+
+        return expandedKeytoolCmd;
+    }
+
+    /**
+     * Constructs the command to pass to keytool for adding the self-signed cert to the trust store
+     *
+     * @param keystore           The target key store that the certificate was added to
+     * @param truststore         The target trust store to add the certificate to
+     * @param keystorePassword   The password for the key store
+     * @param truststorePassword The password for the trust store
+     * @param alias              The alias of the certificate
+     * @return A String array to pass to {@link com.sun.enterprise.admin.servermgmt.KeystoreManager.KeytoolExecutor}
+     */
+    public static String[] constructImportCertKeytoolCommand(File keystore, File truststore, char[] keystorePassword,
+            char[] truststorePassword, String alias) {
+        String[] keytoolCmd = new String[]{"-importkeystore", "-srckeystore", keystore.getAbsolutePath(),
+                "-destkeystore", truststore.getAbsolutePath(), "-srcalias", alias, "-destalias", alias,
+                "-srcstorepass", new String(keystorePassword), "-deststorepass", new String(truststorePassword),
+                "-srckeypass", new String(keystorePassword), "-destkeypass", new String(truststorePassword),
+                "-noprompt"};
+
+        return keytoolCmd;
+    }
+
+    /**
+     * @param parser    The {@link MiniXmlParser} for extracting info from the domain.xml
+     * @param listener  The name of the listener to get the password from.
+     * @param attribute The name of the store password attribute (should be key-store-password or trust-store-password)
+     * @return A char array containing the password of the target listener, or null if no matches or password found
+     * @throws MiniXmlParserException if there's an issue reading the domain.xml
+     */
+    public static char[] getPasswordFromListener(MiniXmlParser parser, String listener, String attribute)
+            throws MiniXmlParserException {
+        char[] password = null;
+        for (Map<String, String> listenerAttributes : parser.getProtocolAttributes()) {
+            if (listenerAttributes.get("name").equals(listener)) {
+                // Get the keystore from the listener if it has a custom one
+                password = listenerAttributes.get(attribute).toCharArray();
```

Comment:
This causes a NPE when listenerAttributes does not have a password attribute which I apparently managed to create as test case. I suspect the same applies to the assignment in 312.